Parse error: Failed to construct DelphiFile (line 5883:0 missing FORWARD at 'var')

### Prerequisites

- [X] This bug is in SonarDelphi, not SonarQube or my Delphi code.
- [X] This bug has not already been reported.

### SonarDelphi version

1.7.0

### SonarQube version

10.5.1

### Issue description

The following error occurs on multiple files in our project. Luckily, one of the files is FastMM5, making it possible to report this bug without sharing our own code. The following file causes the error:

https://github.com/pleriche/FastMM5/blob/version_505/FastMM5.pas

This causes the sonar-delphi to crash with the following stacktrace:

```
14:44:25.820 ERROR Error while processing <projectpath>\fastmm\FastMM5.pas

au.com.integradev.delphi.file.DelphiFile$DelphiFileConstructionException: Failed to construct DelphiFile (line 5883:0 missing FORWARD at 'var')
	at au.com.integradev.delphi.file.DelphiFile.setupFile(DelphiFile.java:161)
	at au.com.integradev.delphi.file.DelphiFile.from(DelphiFile.java:136)
	at au.com.integradev.delphi.symbol.SymbolTableBuilder.process(SymbolTableBuilder.java:284)
	at au.com.integradev.delphi.symbol.SymbolTableBuilder.processImportsWithInlineRoutines(SymbolTableBuilder.java:371)
	at au.com.integradev.delphi.symbol.SymbolTableBuilder.process(SymbolTableBuilder.java:293)
	at au.com.integradev.delphi.symbol.SymbolTableBuilder.indexUnit(SymbolTableBuilder.java:396)
	at au.com.integradev.delphi.symbol.SymbolTableBuilder.build(SymbolTableBuilder.java:477)
	at au.com.integradev.delphi.DelphiSensor.executeOnFiles(DelphiSensor.java:119)
	at au.com.integradev.delphi.DelphiSensor.execute(DelphiSensor.java:83)
	at org.sonar.scanner.sensor.AbstractSensorWrapper.analyse(AbstractSensorWrapper.java:64)
	at org.sonar.scanner.sensor.ModuleSensorsExecutor.execute(ModuleSensorsExecutor.java:88)
	at org.sonar.scanner.sensor.ModuleSensorsExecutor.lambda$execute$1(ModuleSensorsExecutor.java:61)
	at org.sonar.scanner.sensor.ModuleSensorsExecutor.withModuleStrategy(ModuleSensorsExecutor.java:79)
	at org.sonar.scanner.sensor.ModuleSensorsExecutor.execute(ModuleSensorsExecutor.java:61)
	at org.sonar.scanner.scan.SpringModuleScanContainer.doAfterStart(SpringModuleScanContainer.java:82)
	at org.sonar.core.platform.SpringComponentContainer.startComponents(SpringComponentContainer.java:226)
	at org.sonar.core.platform.SpringComponentContainer.execute(SpringComponentContainer.java:205)
	at org.sonar.scanner.scan.SpringProjectScanContainer.scan(SpringProjectScanContainer.java:204)
	at org.sonar.scanner.scan.SpringProjectScanContainer.scanRecursively(SpringProjectScanContainer.java:200)
	at org.sonar.scanner.scan.SpringProjectScanContainer.doAfterStart(SpringProjectScanContainer.java:173)
	at org.sonar.core.platform.SpringComponentContainer.startComponents(SpringComponentContainer.java:226)
	at org.sonar.core.platform.SpringComponentContainer.execute(SpringComponentContainer.java:205)
	at org.sonar.scanner.bootstrap.SpringScannerContainer.doAfterStart(SpringScannerContainer.java:351)
	at org.sonar.core.platform.SpringComponentContainer.startComponents(SpringComponentContainer.java:226)
	at org.sonar.core.platform.SpringComponentContainer.execute(SpringComponentContainer.java:205)
	at org.sonar.scanner.bootstrap.SpringGlobalContainer.doAfterStart(SpringGlobalContainer.java:138)
	at org.sonar.core.platform.SpringComponentContainer.startComponents(SpringComponentContainer.java:226)
	at org.sonar.core.platform.SpringComponentContainer.execute(SpringComponentContainer.java:205)
	at org.sonar.batch.bootstrapper.Batch.doExecute(Batch.java:71)
	at org.sonar.batch.bootstrapper.Batch.execute(Batch.java:65)
	at org.sonarsource.scanner.lib.internal.batch.BatchIsolatedLauncher.execute(BatchIsolatedLauncher.java:41)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at org.sonarsource.scanner.lib.internal.IsolatedLauncherProxy.invoke(IsolatedLauncherProxy.java:62)
	at jdk.proxy3/jdk.proxy3.$Proxy2.execute(Unknown Source)
	at org.sonarsource.scanner.lib.InProcessScannerEngineFacade.doAnalyze(InProcessScannerEngineFacade.java:39)
	at org.sonarsource.scanner.lib.ScannerEngineFacade.analyze(ScannerEngineFacade.java:61)
	at org.sonarsource.scanner.cli.Main.analyze(Main.java:77)
	at org.sonarsource.scanner.cli.Main.main(Main.java:63)
Caused by: au.com.integradev.delphi.antlr.DelphiParser$ParserException: line 5883:0 missing FORWARD at 'var'
	at au.com.integradev.delphi.antlr.DelphiParser.reportError(DelphiParser.java:387)
	at org.antlr.runtime.BaseRecognizer.recoverFromMismatchedToken(BaseRecognizer.java:613)
	at org.antlr.runtime.BaseRecognizer.match(BaseRecognizer.java:115)
	at au.com.integradev.delphi.antlr.DelphiParser.forwardDirectiveSection(DelphiParser.java:32532)
	at au.com.integradev.delphi.antlr.DelphiParser.forwardRoutineHeading(DelphiParser.java:16915)
	at au.com.integradev.delphi.antlr.DelphiParser.forwardRoutine(DelphiParser.java:16097)
	at au.com.integradev.delphi.antlr.DelphiParser.routineImplementation(DelphiParser.java:15863)
	at au.com.integradev.delphi.antlr.DelphiParser.declSection(DelphiParser.java:3480)
	at au.com.integradev.delphi.antlr.DelphiParser.unitImplementation(DelphiParser.java:2002)
	at au.com.integradev.delphi.antlr.DelphiParser.unit(DelphiParser.java:1618)
	at au.com.integradev.delphi.antlr.DelphiParser.file(DelphiParser.java:524)
	at au.com.integradev.delphi.file.DelphiFile.createAST(DelphiFile.java:196)
	at au.com.integradev.delphi.file.DelphiFile.setupFile(DelphiFile.java:149)
	... 38 common frames omitted
Caused by: org.antlr.runtime.MissingTokenException: null
	at org.antlr.runtime.BaseRecognizer.recoverFromMismatchedToken(BaseRecognizer.java:612)
	... 49 common frames omitted
```

### Steps to reproduce

The easiest way to reproduce this is to do a DelphiLint analysis on the following file:

https://github.com/pleriche/FastMM5/blob/version_505/FastMM5.pas

This shows the same error as sonar-delphi.

### Minimal Delphi code exhibiting the issue

https://github.com/pleriche/FastMM5/blob/version_505/FastMM5.pas

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Fixed
+
+- Parsing errors around partial `asm` blocks blocks in conditional branches.
+
 ## [1.8.0] - 2024-08-02
 
 ### Added

--- a/delphi-frontend/src/main/antlr3/au/com/integradev/delphi/antlr/Delphi.g
+++ b/delphi-frontend/src/main/antlr3/au/com/integradev/delphi/antlr/Delphi.g
@@ -1201,7 +1201,16 @@ ASSEMBLER         : A S S E M B L E R           ;
 ASSEMBLY          : A S S E M B L Y             ;
 AT                : A T                         ;
 AUTOMATED         : A U T O M A T E D           ;
-BEGIN             : B E G I N                   ;
+BEGIN             : B E G I N {
+                    // HACK:
+                    //  We exit asmMode in the lexer if we encounter a BEGIN token, which is a hack
+                    //  to support partial asm blocks sharing an END keyword with some other block
+                    //  due to conditional compilation.
+                    //
+                    //  For more information, see:
+                    //    https://github.com/integrated-application-development/sonar-delphi/issues/116
+                    asmMode = false;
+                  };
 CASE              : C A S E                     ;
 CDECL             : C D E C L                   ;
 CLASS             : C L A S S                   ;

--- a/delphi-frontend/src/test/java/au/com/integradev/delphi/antlr/GrammarTest.java
+++ b/delphi-frontend/src/test/java/au/com/integradev/delphi/antlr/GrammarTest.java
@@ -343,4 +343,9 @@ class GrammarTest {
   void testImmediatelyTerminatedComments() {
     assertParsed("ImmediatelyTerminatedComments.pas");
   }
+
+  @Test
+  void testConditionalAsm() {
+    assertParsed("ConditionalAsm.pas");
+  }
 }

--- a/delphi-frontend/src/test/resources/au/com/integradev/delphi/grammar/ConditionalAsm.pas
+++ b/delphi-frontend/src/test/resources/au/com/integradev/delphi/grammar/ConditionalAsm.pas
@@ -1,0 +1,16 @@
+unit ConditionalAsm;
+
+interface
+
+implementation
+
+procedure Foo;
+{$if False}
+asm
+{$else}
+begin
+  var x := @Foo;
+{$endif}
+end;
+
+end.

